### PR TITLE
CBL-1680: Phase 1, Add JSON API

### PR DESCRIPTION
**Pull request by bmeike**
I believe this is the complete JSON API as spec'ed in https://docs.google.com/document/d/1H0mnutn-XXIADvGT_EjINAOVwt0Ea8vwW70v0i_PO54

There is no implementation here.  This is just the API and fixed commenting.

Note that, although not specifically mentioned in the spec, I've updated the interfaces that define the Dictionary and Array APIs

### CODE REVIEW

File: common/main/java/com/couchbase/lite/MutableArray.java [hunk 1]
```diff
@@ -38,41 +38,45 @@
...
     /**
-     * Constructs a new Array object with an array content. Allowed value types are List, Date,
-     * Map, Number, null, String, Array, Blob, and Dictionary. The List and Map must contain
-     * only the above types.
+     * Creates a new MutableArray with content from the passed JSON string.
      *
-     * @param data the array object.
+     * @param json the array content as a JSON string.
      */
-    public MutableArray(List<Object> data) {
```
Comment (Sandychuang8) [inline common/main/java/com/couchbase/lite/MutableArray.java:52]: Should we remove this constructor??

File: common/main/java/com/couchbase/lite/MutableArray.java [hunk 2]
```diff
@@ -82,11 +86,26 @@ public MutableArray(List<Object> data) {
...
+     * Allowed value types are List, Date, Map, Number, null, String, Array, Blob, and Dictionary.
+     * The List and Map must contain only the above types. Setting the new dictionary content will
+     * replace the current data including the existing Array and Dictionary objects.
+     *
+     * @param json the dictionary object.
+     * @return this Document instance
+     */
+    @NonNull
+    @Override
+    public MutableArray setJSON(@NonNull String json) {
```
Comment (Sandychuang8) [inline common/main/java/com/couchbase/lite/MutableArray.java:105]: Not ToJSON? Is this because you are following java naming convention?

**Review by Sandychuang8 — APPROVED** [review]

File: common/main/java/com/couchbase/lite/MutableArray.java [hunk 3]
```diff
@@ -38,41 +38,45 @@
...
     //---------------------------------------------
     // API - public methods
     //---------------------------------------------
 
     /**
-     * Set an array as a content. Allowed value types are List, Date,
-     * Map, Number, null, String, Array, Blob, and Dictionary. The List and Map must contain
-     * only the above types. Setting the new array content will replace the current data
-     * including the existing Array and Dictionary objects.
+     * Populate an array with content from a Map.
```
Comment (jayahariv) [inline common/main/java/com/couchbase/lite/MutableArray.java:76]: confused! you mean Map or List? Argument is `List<Object>`?

**Review by jayahariv — APPROVED** [review]

File: common/main/java/com/couchbase/lite/MutableArray.java [hunk 4]
```diff
@@ -82,11 +86,26 @@ public MutableArray(List<Object> data) {
     public MutableArray setData(List<Object> data) {
         synchronized (lock) {
             internalArray.clear();
-            for (Object obj : data) { internalArray.append(Fleece.toCBLObject(obj)); }
+            for (Object obj: data) { internalArray.append(Fleece.toCBLObject(obj)); }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/MutableArray.java:89]: Can you add @NonNull to data param as well?

**Review by pasin — APPROVED** [review]

File: common/main/java/com/couchbase/lite/MutableArray.java (hunk 3 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/MutableArray.java:76]: Yeah... it is not totally clear.  The idea is that the List can contains Maps or Lists.  If it does, those Maps/Lists must contain only the types mentioned above.  I've revisited the comments to try to make them clearer.

File: common/main/java/com/couchbase/lite/MutableArray.java (hunk 2 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/MutableArray.java:105]: setJSON() is in the MutableArray.  toJSON() is inherited from (immutable) Array.

Is that how you did it?

File: common/main/java/com/couchbase/lite/MutableArray.java (hunk 1 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/MutableArray.java:52]: Yeah... I don't think the spec suggests removing this.  It seems consistent with the constructors for Document and Dictionary objects.

File: common/main/java/com/couchbase/lite/MutableArray.java (hunk 4 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/MutableArray.java:89]: Yes.  Definitely

**Pull request merged**